### PR TITLE
Add SkyPilot Benchmark doc

### PR DESCRIPTION
**Pull request by WoosukKwon**

### CODE REVIEW

File: docs/source/reference/benchmark/index.rst [hunk 1]
```diff
@@ -0,0 +1,34 @@
+.. _benchmark-overview:
+
+Benchmark
+================================================
+
+SkyPilot supports **easy benchmarking of cloud resources**.
+With minimal effort, you can find the right cloud resource for your task and **save significant cost**.
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/index.rst:7]: On reading this, my first thought is "what is benchmarking?" Maybe consider writing as:

```
SkyPilot allows **easy measurement of performance and cost of different kinds of cloud resources** through the benchmark feature.
With minimal effort, you can find the right cloud resource for your task that fits your performance goals and budget constraints.
```

File: docs/source/reference/benchmark/index.rst [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+Here is an example task that fine-tunes BERT on a question answering dataset.
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/index.rst:9]: Consider:
```
For example, say you want to fine-tune a BERT model and you do not know which GPU type is the best for you.
```

File: docs/source/reference/benchmark/index.rst [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    Legend:
+    - STEPS: Number of steps taken.
+    - SEC/STEP, $/STEP: Average time (cost) per step.
+    - EST(hr), EST($): Estimated total time (cost) to complete the benchmark.
+
+    CLUSTER            RESOURCES                                 STATUS      DURATION  SPENT($)  STEPS   SEC/STEP  $/STEP    EST(hr)  EST($)  
+    sky-bench-squad-0  1x Azure(Standard_NC6_Promo, {'K80': 1})  TERMINATED  12m 48s   0.0384    1415    1.1548    0.000058  10.60    1.91    
+    sky-bench-squad-1  1x AWS(g4dn.xlarge, {'T4': 1})            TERMINATED  14m 2s    0.1230    2387    0.6429    0.000094  5.92     3.11    
+    sky-bench-squad-2  1x AWS(g5.xlarge, {'A10G': 1})            TERMINATED  13m 57s   0.2339    7423    0.1859    0.000052  1.75     1.76
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/index.rst:22]: Can we rename `squad` to `bert`? e.g. `sky-bench-bert-0`

File: docs/source/reference/benchmark/index.rst [hunk 4]
```diff
@@ -0,0 +1,34 @@
+.. _benchmark-overview:
+
+Benchmark
+================================================
+
+SkyPilot supports **easy benchmarking of cloud resources**.
+With minimal effort, you can find the right cloud resource for your task and **save significant cost**.
+
+Here is an example task that fine-tunes BERT on a question answering dataset.
+With SkyPilot Benchmark, you can quickly run your task on different types of VMs and get a simple benchmark report like the following:
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/index.rst:10]: remove `simple`

File: docs/source/reference/benchmark/index.rst [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    sky-bench-squad-0  1x Azure(Standard_NC6_Promo, {'K80': 1})  TERMINATED  12m 48s   0.0384    1415    1.1548    0.000058  10.60    1.91    
+    sky-bench-squad-1  1x AWS(g4dn.xlarge, {'T4': 1})            TERMINATED  14m 2s    0.1230    2387    0.6429    0.000094  5.92     3.11    
+    sky-bench-squad-2  1x AWS(g5.xlarge, {'A10G': 1})            TERMINATED  13m 57s   0.2339    7423    0.1859    0.000052  1.75     1.76      
+    sky-bench-squad-3  1x GCP(n1-highmem-8, {'V100': 1})         TERMINATED  13m 45s   0.6768    7306    0.2005    0.000165  1.87     5.51
+
+The report shows the benchmarking results of 4 VMs each with a different GPU type.
+Based on the report, you can pick the VM with either the lowest cost (``EST($)``) or the fastest execution time (``EST(hr)``), or find a sweet spot between them.
+In this example, AWS g5.xlarge (NVIDIA A10G GPU) turns out to be the best choice in terms of both cost and time.
+
+Table of Contents
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/index.rst:29]: I really like the intro till here! However after this point, it seems a little abrupt to simply leave the reader with the CLI, YAML and SkyCallback reference. 

Can we add a paragraph or two about how users can use sky benchmark? I would suggest something like:

```
# Using SkyBench
A part of the Sky Benchmark report relies on the `SkyCallback` library <link> instrumented in the training code to report step completion. Depending on the level of detail required by you in the benchmark report, Sky Benchmark can be used in two modes:

1. Without SkyCallback - <explain or link to CLI page>. **This requires zero changes in your code**.
2. With SkyCallback - <explain or link to SkyCallback page>.
```
^This above content is just a draft, please feel free to refine it.

File: docs/source/reference/benchmark/callback.rst [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            total_steps=num_epochs * len(train_dataloader), # Optional
+        )
+
+Integrations with ML frameworks
+----------------------------------------------------------
+
+Using SkyCallback is even easier for **Keras, PytorchLightning, and HuggingFace Transformers** programs where trainer APIs are used.
+SkyCallback natively supports these frameworks with simple interface.
+
+* Keras example
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/callback.rst:88]: Can we also link to small examples either in our repository or elsewhere?

**Review by romilbhardwaj — COMMENTED** [review]
This is great, thanks for adding @WoosukKwon! Viewed index.rst and left some comments

**Comment by WoosukKwon** [comment]
Thanks for your review @romilbhardwaj . I've addressed your comments. PTAL.

File: docs/source/reference/benchmark/index.rst [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+    CLUSTER            RESOURCES                                STATUS      DURATION  SPENT($)  STEPS   SEC/STEP  $/STEP    EST(hr)  EST($)  
+    sky-bench-bert-0  1x Azure(Standard_NC6_Promo, {'K80': 1})  TERMINATED  12m 48s   0.0384    1415    1.1548    0.000058  10.60    1.91    
+    sky-bench-bert-1  1x AWS(g4dn.xlarge, {'T4': 1})            TERMINATED  14m 2s    0.1230    2387    0.6429    0.000094  5.92     3.11    
+    sky-bench-bert-2  1x AWS(g5.xlarge, {'A10G': 1})            TERMINATED  13m 57s   0.2339    7423    0.1859    0.000052  1.75     1.76      
+    sky-bench-bert-3  1x GCP(n1-highmem-8, {'V100': 1})         TERMINATED  13m 45s   0.6768    7306    0.2005    0.000165  1.87     5.51
+
+The report shows the benchmarking results of 4 VMs each with a different GPU type.
+Based on the report, you can pick the VM with either the lowest cost (``EST($)``) or the fastest execution time (``EST(hr)``), or find a sweet spot between them.
+In this example, AWS g5.xlarge (NVIDIA A10G GPU) turns out to be the best choice in terms of both cost and time.
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/index.rst:27]: Is it possible to have an example which does not have a obvious best choice - e.g., one with a fast but expensive GPU and one with a slow but overall cheaper GPU? This is not necessary, but great to have

File: docs/source/reference/benchmark/index.rst [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+Using SkyPilot Benchmark
+------------------------
+
+A part of the SkyPilot Benchmark report relies on the :ref:`SkyCallback` library instrumented in the training code to report step completion.
+Depending on the level of detail required by you in the benchmark report, SkyPilot Benchmark can be used in two modes:
+
+1. Without SkyCallback - You can get a basic benchmark report using SkyPilot Benchmark :ref:`benchmark-cli`. **This requires zero changes in your code**.
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/index.rst:35]: Add `Please refer to <cli page>.`  to the end?

File: docs/source/reference/benchmark/cli.rst [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+Workflow
+--------------------------------
+
+You can use SkyPilot Benchmark by simply replacing your ``sky launch`` command with ``sky bench launch``:
+
+.. code-block:: bash
+
+    # Normal launch command
+    $ sky launch mytask.yaml --gpus V100
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/cli.rst:14]: Maybe we should remove this to make the code block more copy paste friendly

File: docs/source/reference/benchmark/callback.rst [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+===========
+
+SkyCallback is a simple Python library that works in conjunction with SkyPilot Benchmark.
+It enables SkyPilot to provide a more detailed benchmark report without the need to wait until the task finishes.
+
+What SkyCallback is for
+--------------------------------------------
+
+SkyCallback is designed for **machine learning tasks** which have a loop iterating many `steps`.
+SkyCallback measures the average time taken by each step, and extrapolates it to the total execution time of the task.
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/callback.rst:13]: How is total execution time extrapolated? Do we know the total number of steps? Good to mention here

File: docs/source/reference/benchmark/callback.rst [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+
+Using SkyCallback generic APIs
+--------------------------------------------
+
+The SkyCallback generic APIs are for **PyTorch, TensorFlow, and JAX** programs where training loops are exposed to the users.
+Below we provide the instructions for using the APIs.
+
+First, import the SkyCallback package and initialize it using ``init``.
+You can optionally specify in ``total_steps`` the total number of steps that the task will iterate through.
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/callback.rst:36]: What happens if you specify it? What happens if you dont specify it?

File: docs/source/reference/benchmark/callback.rst [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+* Keras example
+
+.. code-block:: python
+
+    from sky_callback import SkyKerasCallback
+
+    # Add the callback to your Keras model.
+    model.fit(..., callbacks=[SkyKerasCallback()])
+
+`Here <https://github.com/skypilot-org/skypilot/blob/master/examples/benchmark/keras_asr/callback.patch>`_ you can find an example of applying SkyCallback to Keras ASR model training.
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/callback.rst:101]: Git patches, while convenient to use, are a little hard to read. Is there some other way we can show this code? Maybe a link to a fork with the patch applied? What do you think?

File: docs/source/reference/benchmark/callback.rst [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    trainer = transformers.Trainer(..., callbacks=[SkyTransformersCallback()])
+
+`Here <https://github.com/skypilot-org/skypilot/blob/master/examples/benchmark/transformers_qa/callback.patch>`_ you can find an example of applying SkyCallback to HuggingFace BERT fine-tuning.
+
+.. note::
+    When using the framework-integrated callbacks, do not call ``sky_callback.init`` for initialization.
+    The callbacks will do it for you.
+
+.. SkyCallback for containerized jobs (Advanced)
+.. --------------------------------------------
```
Comment (romilbhardwaj) [inline docs/source/reference/benchmark/callback.rst:130]: Mark as TODO?

**Review by romilbhardwaj — COMMENTED** [review]
Great job @WoosukKwon! 

My review covers only high level content and making sure all required information is here. I have refrained from making nit/structuring comments in this review because I believe those are better fixed in a separate PR - I can create that PR after this is merged and you can review it.

File: docs/source/reference/benchmark/index.rst (hunk 8 above)
Comment (WoosukKwon) [reply docs/source/reference/benchmark/index.rst:35]: :ref:`benchmark-cli` is the link to the CLI page. Do you mean adding another one at the end?

File: docs/source/reference/benchmark/callback.rst (hunk 12 above)
Comment (WoosukKwon) [reply docs/source/reference/benchmark/callback.rst:101]: I thought showing the full training code is inefficient as we need only two lines of code changes. I believe most users will understand the APIs without seeing the examples, and the examples are only to confirm that no other changes are required.

File: docs/source/reference/benchmark/cli.rst (hunk 9 above)
Comment (WoosukKwon) [reply docs/source/reference/benchmark/cli.rst:14]: Good point. Users of this feature will be familiar to SkyPilot basic APIs, so we can skip that.

File: docs/source/reference/benchmark/callback.rst (hunk 13 above)
Comment (WoosukKwon) [reply docs/source/reference/benchmark/callback.rst:130]: Removed. I will add this to the doc when we have a user need.

**Review by romilbhardwaj — APPROVED** [review]

**Pull request merged**